test/test_pattern.py

```diff
@@ -0,0 +1,42 @@
+from __future__ import absolute_import, division, print_function
+
+from funsor.domains import reals
+from funsor.interpreter import interpretation, reinterpret
+from funsor.terms import Number, Variable, lazy
+
+from funsor.pattern import match, match_vars, unify_interpreter
+from unification import unify
+
+
+def test_unify_binary():
+    with interpretation(lazy):
+        pattern = Variable('a', reals()) + Number(2.) * Variable('b', reals())
+        expr = Number(1.) + Number(2.) * (Number(3.) - Number(4.))
+
+    subs = unify(pattern, expr)
+    print(subs, pattern(**{k.name: v for k, v in subs.items()}))
+    assert subs is not False
+
+    with interpretation(unify_interpreter):
+        assert unify((pattern,), (expr,)) is not False
+
+
+def test_match_binary():
+    with interpretation(lazy):
+        pattern = Variable('a', reals()) + Number(2.) * Variable('b', reals())
+        expr = Number(1.) + Number(2.) * (Number(3.) - Number(4.))
+
+    @match_vars(pattern)
+    def expand_2_vars(a, b):
+        return a + b + b
```

Comment:
This shows how `match_vars` will be helpful for interpreters that use fancier patterns